src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -152,188 +164,159 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
+                        {
+                            // Sequence needs to be CRLF and not contain an inner CR not part of terminator.
+                            // Not parsable as a valid name:value header pair.
+                            RejectRequestHeader(span[..lineLength]);
                         }
+
+                        // Read the header successfully, skip the reader forward past the headerSpan.
+                        span = span[lineLength..];
+                        reader.Advance(lineLength);
                     }
 
                     // End found in current span
                     if (length > 0)
                     {
+                        length = -1;
                         continue;
                     }
 
-                    // We moved the reader to look ahead 2 bytes so rewind the reader
-                    if (readAhead > 0)
-                    {
-                        reader.Rewind(readAhead);
-                    }
+                    // Load next header line to parse as a span
+                    span = ParseMultiSpanHeader(ref reader, out length);
 
-                    length = ParseMultiSpanHeader(handler, ref reader);
-                    if (length < 0)
+                    // If there any remaining line?
+                    if (length == -1 && span.Length == 0)
                     {
-                        // Not there
                         return false;
                     }
-
-                    reader.Advance(length);
-                    // As we crossed spans set the current span to default
-                    // so we move to the next span on the next iteration
-                    span = default;
                 }
             }
 
             return false;
         }
 
-        private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byte> reader)
+        // Returns the length of the line terminator (CRLF = 2, LF = 1)
+        // If no valid EOL is detected then -1
+        private int ParseHeaderLineEnd(ReadOnlySpan<byte> headerSpan, int length)
+        {
+            if (headerSpan[length] == ByteCR)
+            {
+                // No more chars after CR? Don't consume an incomplete header
+                if (headerSpan.Length == length + 1)
+                {
+                    return 0;
+                }
+
+                // CR must be followed by LF in all cases
+                if (headerSpan[length + 1] != ByteLF)
+                {
+                    if (length == 0)
+                    {
+                        KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
+                    }
+                    else
+                    {
+                        RejectRequestHeader(headerSpan[..(length + 2)]);
+                    }
+                }
+
+                return 2;
+            }
+
+            if (_enableLineFeedTerminator)
+            {
+                return 1;
```

Comment:
The assumption here is that the caller used IndexOfAny(CR,LF) to pass in the length. This works but I'd feel better if that assumption was made more explicit so we don't accidentally mess that up in the future.